{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 200926\n",
    "1. UniRep github\n",
    "2. Active Matter paper by Tyler\n",
    "3. Background on motor and microtubule\n",
    "    1) microtubules are largest components of the cytoskeleton, and are polymers of tubulin dimers\n",
    "    2) centrosome is the main microtubule organizing center of animal cells composed of two centrioles\n",
    "    3) Motor proteins hydrolyze ATP and cause conformational change that produce movement. Example: kinesin transports vesicles along microtubule tracks\n",
    "    4) Small roads are Actin, main road microtubule\n",
    "    - Julie Theriot: Protein Polymers, Crawling Cells, and Cornet Tails\n",
    "        - cell organizmation is dynamic: neutrophil chasing a bacteria, amoebae search for food (separated by 2 billion years of evolution)\n",
    "        - dense network of cross-linked and branched filaments.\n",
    "        - polymerization of actin/tubulin could generate physical force\n",
    "        - actin polymerization optical trap from horseshoe crab sperm acrosome\n",
    "        - Listeria monocytogenes: actin-based motility --> Movement in cytoplasmic extract replace host cell ---> **ActA-coated** polystyrene beads replace bacterial\n",
    "    - Julie Theriot: Mechanics and Dynamics of Rapid Cell Motility\n",
    "        - at the leading edge of motile cell: assembly of actin as a nanomachine fish epidermal cell - keratocytes (F-actin polymerization, myosinII contraction,adhesive forces by integrins)\n",
    "        - shape of the cell determined by how fast the actiin is polymerizing\n",
    "        - can sense and respond to electric fields\n",
    "        - neutrophil actomyosin mechanism\n",
    "   \n",
    "   - Julie Theriot: Evolution of a Dynamic Cytoskeleton\n",
    "       - LUCA last universal common ancestor : similarity of ribosomal RNA\n",
    "       - Archaea and bateria and no long similar than they are to eukaryotes but their morphology are similar\n",
    "       - bacteria: even the largest and most complicated have diffusion limited dimension\n",
    "       - Bacteria actually have tubulin FtsZ true homologs\n",
    "       - Bacteria have actin homolog (Mbl an MreB) also in magnetosome\n",
    "       - Larg-scale cell organization by all cytoskeletons requires dynamic assembly and disassembly of helical protein filamens\n",
    "       - Hypothesis: Eukaryotes enhance these features with specialize cytoskeleton-associated factors\n",
    "       - structures not requiring motors or nucleators: single filament, mixed-polarity bundle\n",
    "       - structures requiring localized nucleation/and or motor: Aster, Parallel bundle, spindle\n",
    "       - microtubules, actin, intermediate filaments\n",
    "       - tubulin-GTP, actin-ATP (binded polymerize, detached deassembly)\n",
    "       - **actin VS hexokinase (structurally similar)**\n",
    "       - Caulobacter crescentus: MreB = Actin FtsZ = Tubulin CreS = Intermediate Filament\n",
    "       - **Exceptions: Stella humosa (star shaped), Haloquadratu walsbyii (square), Epulopiscium fishelsonii (HUGE)**\n",
    "       - Nucleators: specialized subunit --> diverge a bit \n",
    "       - Motor: Myosins, Kinesins, Dyneins\n",
    "       - P-loop NTPases ONLY in Eukaryotes (other ~50 subclasses)\n",
    "       - bacterial flagella rotor and type IV pili (nothing seems to do things inside)\n",
    "   - Ron Vale (UCSF, HHMI) 1: Molecular Motor Proteins\n",
    "       - Dynein: towards minus ends, Kinesin to + end\n",
    "       - Myosin on actin filaments\n",
    "       - Kinesin I Domain Organization Motor--Coiled coild--Hinge--Tail(cargo)\n",
    "       - **Measurements** In Vitro Motility Assays, Optical Trap to measure stall Force, measure distance of steps through fluorophore\n",
    "       - Myosin and Kinesin has very similar hydrolyzation core and relay helix structures (mechanical elements that diverge through evolution)\n",
    "       - Kinesin 14 towards - end motor: a different mechanical elements\n",
    "       - Diseases: Familial hypertrophic cardiomyopathy (cardiac myosin), deafness (Myosin VI, myosin VII), Cilia Dyskinesias (Dynein and Associated Proteins), Neurodegenerative diseases (kinesin motors)\n",
    "   - Ron Vale (UCSF, HHMI) 2: Molecular Motor Proteins: The Mechanism of Dynein Motility\n",
    "       - 8 times larger than kinesin motor domain\n",
    "       - Axonemal Dynein: sinodual doublet microtubules\n",
    "       - Cytoplasmic dynein I : - end directed cargo transport (towards the center)\n",
    "       - Myosin and Kinesin more phylogenetically related than Dynein (AAA ATPase)\n",
    "   - Ron Vale (UCSF, HHMI) 3: Molecular Motor Proteins: Regulation of Mammalian Dynein\n",
    "       - Bicaudal-D (BicD),Dynactin,Dynein: One of the most processive motor through the complex"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Big Aims:\n",
    "    1. RNN embedding for the Eukaryotic molecular motors and apply the embedding to see if phylogenetic/functional/domain-specific/mechanism-specific could be learned\n",
    "    2. Top Models could be built upon characterized motors with labels on their in vitro measurements\n",
    "    3. larger comparison of embedding from Prokaryotes and Archae to mine comparatives of motor proteins/study of those special case prokaryotes on embedded protein space"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Questions\n",
    "- each domain as a training sequence to enable character VS word level learning\n",
    "- the motor proteins have much longer sequence than 275 limit for publicly available training of UniRep\n",
    "- The scope of motor proteins included (P-loop NTPases)? The redundency of the sequence is much higher than UniRef50\n",
    "- https://www.cymobase.org/cymobase [Cymobase: A database for cytoskeletal and motor protein sequence information]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 200928\n",
    "1. Data File Selection PFam:\n",
    "    - from the phylogenetic tree, I narrowed down to the pFam superfamily (clan) of P-loop NTPase\n",
    "    -     \n",
    "    `\n",
    "    #=GF ID   P-loop_NTPase\n",
    "#=GF AC   CL0023.34\n",
    "#=GF PI   AAA;\n",
    "#=GF DE   P-loop containing nucleoside triphosphate hydrolase superfamily\n",
    "    `\n",
    "    - Could download the sequences corresponding to the families in the clan via Pfam-A.fasta.gz\n",
    "    - mappipng \n",
    "2. Data selection UniRef50:\n",
    "    - Could search words such as \"kinesin\", but clusters could not be directly related to pfam families\n",
    "    - could alternatively search iteratively against the Uniprot database using the family names from pfam\n",
    "\n",
    "3. What could be done by UniRep:\n",
    "    - weights of the mLSTM could be loaded and then train a top model with x and y values. If we only want the embedded vectors, we should get the final_hidden variable frome the model \n",
    "\n",
    "4. What could be done by TAPE from Yun Song lab:\n",
    "    - Benchmarking datasets for 5 different tasks\n",
    "    - could easily get embedding vectors from 3 different pretrained models (Transformer, UniRep1900, trRosetta based on CNN ResNet of MSA)\n",
    "\n",
    "5. What SeqVec could do from Burkhard Rost?:\n",
    "    - a novel way to represent protein sequences as continuous vectors (embeddings) by using the language model ELMo taken from natural language processing. \n",
    "    - pretrained model on UniRef50\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dear Prof.Thomson,\n",
    "    As I tried to familiarize myself with potential embedding models and training datasets for molecular motor last week, I would like to discuss with you on the next step of the project.\n",
    "    \n",
    "    As for the set of molecular motors to be embedded, I currently found the following sources that could be helpful:\n",
    "    1. From Pfam, I could get motor sequences and their evolutionary related proteins grouped in P-Loop NTPase Clan (superfamily)\n",
    "    2. Since most of the pre-trained embedding models are based on UniRef50 dataset, I could extract clusters with name containing words such as \"kinesin\" from UniRef50. However, I currently haven't found an effective way to link UniRef50 and Pfam, whereas I could do with UniprotKB and Pfam.\n",
    "    3. From [Cymobase: A database for cytoskeletal and motor protein sequence information]. This is a database with curated information and proteins properties that could potentially use to label our data. However, it seems to only offer server-based query \n",
    "    \n",
    "    As for the embedding models, I found the following published pre-trained NLP-driven models that could directly generate sequence embeddings:\n",
    "    1. UniRep is a unidirectional mLSTM with 1900 units\n",
    "    2. TAPE from Yun Song lab at UC Berkeley provide pretrain models in pytorch architecture, including BERT-based Transformer,UniRep, and trRosetta (cnn ResNet on MSA)\n",
    "    3. SeqVec from Burkhard Rost lab trained bi-directional EMLo on UniRef50\n",
    "\n",
    "    To my current knowledge, people tune these pre-trained models mostly using UniRef50 dataset due to its non-redundency, and I'm curious if these models therefore capture more generalized features and may not distinguish well for a specific class such as molecular motors. However, from the GFP engineering example in UniRep, it seems promising to capture structural and functional features of a specific family of proteins. \n",
    "    \n",
    "    With the above reasoning, I would :\n",
    "    1. try to build up embedding vectors for the P-Loop NTPase superfamily from pfam using the 4 pre-trained models above. \n",
    "    2. try later to provide motor protein properties (in vitro motility, optical trap, phylogenetic distance, and moving step size of motors) as labels to either inspect the embedded space or to train supervised top models. \n",
    "    3. try see if prokaryotes with abnormal shape and size may have structurely/functionally similar protein in the embedded space. Some Examples: Stella humosa (star shaped), Haloquadratu walsbyii (square), Epulopiscium fishelsonii (HUGE)\n",
    "    \n",
    "    The proposed plan may also change since I'm currently unsure if my local computing environment could handle the tasks. So far I've only explored molecular motors and I'm wondering if and how we would like to include tubulin and actin into the space, their homologs exists in prokaryotes (in Caulobacter crescentus: MreB = Actin FtsZ = Tubulin CreS = Intermediate Filament).\n",
    "    It would be great if you could connect me to lab members that have done previous exploratory tasks on this project, and if there're any computing cluster resource available for my potential usage.\n",
    "    Best,\n",
    "    Changhua\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 200930\n",
    "1. Get the Pfams of P-loop NTPase\n",
    "    - \n",
    "    `\n",
    "    grep -A 500 'CL0023.34' Pfam-C\n",
    "    `\n",
    "    - \n",
    "    `\n",
    "    #=GF AC   CL0023.34\n",
    "#=GF PI   AAA;\n",
    "#=GF DE   P-loop containing nucleoside triphosphate hydrolase superfamily\n",
    "#=GF AU   Studholme DJ\n",
    "#=GF RN   [1]\n",
    "#=GF RM   7646486\n",
    "#=GF RT   A 200-amino acid ATPase module in search of a basic function. \n",
    "#=GF RA   Confalonieri F, Duguet M; \n",
    "#=GF RL   Bioessays 1995;17:639-650.\n",
    "#=GF RN   [2]\n",
    "#=GF RM   9927482\n",
    "#=GF RT   AAA+: A class of chaperone-like ATPases associated with the assembly,\n",
    "#=GF RT   operation, and disassembly of protein complexes. \n",
    "#=GF RA   Neuwald AF, Aravind L, Spouge JL, Koonin EV; \n",
    "#=GF RL   Genome Res 1999;9:27-43.\n",
    "#=GF RN   [3]\n",
    "#=GF RM   11916378\n",
    "#=GF RT   Classification and evolution of P-loop GTPases and related ATPases.\n",
    "#=GF RA   Leipe DD, Wolf YI, Koonin EV, Aravind L;\n",
    "#=GF RL   J Mol Biol. 2002;317:41-72.\n",
    "#=GF RN   [4]\n",
    "#=GF RM   15381417\n",
    "#=GF RT   STAND, a class of P-loop NTPases including animal and plant\n",
    "#=GF RT   regulators of programmed cell death: multiple, complex domain\n",
    "#=GF RT   architectures, unusual phyletic patterns, and evolution by horizontal\n",
    "#=GF RT   gene transfer.\n",
    "#=GF RA   Leipe DD, Koonin EV, Aravind L;\n",
    "#=GF RL   J Mol Biol. 2004;343:1-28.\n",
    "#=GF RN   [5]\n",
    "#=GF RM   14568537\n",
    "#=GF RT   Evolution and classification of P-loop kinases and related proteins.\n",
    "#=GF RA   Leipe DD, Koonin EV, Aravind L;\n",
    "#=GF RL   J Mol Biol. 2003;333:781-815.\n",
    "#=GF DR   CATH; 3.40.50.300;\n",
    "#=GF DR   SCOP; 52540;\n",
    "#=GF CC   AAA family proteins often perform chaperone-like functions that\n",
    "#=GF CC   assist in the assembly, operation, or disassembly of protein\n",
    "#=GF CC   complexes [2].\n",
    "#=GF MB   PF06414;\n",
    "#=GF MB   PF06564;\n",
    "#=GF MB   PF07015;\n",
    "#=GF MB   PF02367;\n",
    "#=GF MB   PF02534;\n",
    "#=GF MB   PF06309;\n",
    "#=GF MB   PF05621;\n",
    "#=GF MB   PF00265;\n",
    "#=GF MB   PF06068;\n",
    "#=GF MB   PF02223;\n",
    "#=GF MB   PF00685;\n",
    "#=GF MB   PF00448;\n",
    "#=GF MB   PF02463;\n",
    "#=GF MB   PF01202;\n",
    "#=GF MB   PF00158;\n",
    "#=GF MB   PF10443;\n",
    "#=GF MB   PF03215;\n",
    "#=GF MB   PF00485;\n",
    "#=GF MB   PF00519;\n",
    "#=GF MB   PF06431;\n",
    "#=GF MB   PF01057;\n",
    "#=GF MB   PF10609;\n",
    "#=GF MB   PF00931;\n",
    "#=GF MB   PF05729;\n",
    "#=GF MB   PF00488;\n",
    "#=GF MB   PF03205;\n",
    "#=GF MB   PF09140;\n",
    "#=GF MB   PF01078;\n",
    "#=GF MB   PF00493;\n",
    "#=GF MB   PF08433;\n",
    "#=GF MB   PF07693;\n",
    "#=GF MB   PF01695;\n",
    "#=GF MB   PF01745;\n",
    "#=GF MB   PF01715;\n",
    "#=GF MB   PF00693;\n",
    "#=GF MB   PF00625;\n",
    "#=GF MB   PF00437;\n",
    "#=GF MB   PF01580;\n",
    "#=GF MB   PF00142;\n",
    "#=GF MB   PF01935;\n",
    "#=GF MB   PF05872;\n",
    "#=GF MB   PF05673;\n",
    "#=GF MB   PF01712;\n",
    "#=GF MB   PF06144;\n",
    "#=GF MB   PF02224;\n",
    "#=GF MB   PF06418;\n",
    "#=GF MB   PF07931;\n",
    "#=GF MB   PF02492;\n",
    "#=GF MB   PF01121;\n",
    "#=GF MB   PF01656;\n",
    "#=GF MB   PF00308;\n",
    "#=GF MB   PF03668;\n",
    "#=GF MB   PF00006;\n",
    "#=GF MB   PF02374;\n",
    "#=GF MB   PF03308;\n",
    "#=GF MB   PF01637;\n",
    "#=GF MB   PF01583;\n",
    "#=GF MB   PF03969;\n",
    "#=GF MB   PF00406;\n",
    "#=GF MB   PF00709;\n",
    "#=GF MB   PF00005;\n",
    "#=GF MB   PF08298;\n",
    "#=GF MB   PF07728;\n",
    "#=GF MB   PF07726;\n",
    "#=GF MB   PF07724;\n",
    "#=GF MB   PF00004;\n",
    "#=GF MB   PF05707;\n",
    "#=GF MB   PF11496;\n",
    "#=GF MB   PF10649;\n",
    "#=GF MB   PF10412;\n",
    "#=GF MB   PF10236;\n",
    "#=GF MB   PF09820;\n",
    "#=GF MB   PF09818;\n",
    "#=GF MB   PF09547;\n",
    "#=GF MB   PF09439;\n",
    "#=GF MB   PF09037;\n",
    "#=GF MB   PF08423;\n",
    "#=GF MB   PF07755;\n",
    "#=GF MB   PF07652;\n",
    "#=GF MB   PF07088;\n",
    "#=GF MB   PF06990;\n",
    "#=GF MB   PF06733;\n",
    "#=GF MB   PF05970;\n",
    "#=GF MB   PF05894;\n",
    "#=GF MB   PF05879;\n",
    "#=GF MB   PF05876;\n",
    "#=GF MB   PF05783;\n",
    "#=GF MB   PF05127;\n",
    "#=GF MB   PF04670;\n",
    "#=GF MB   PF04466;\n",
    "#=GF MB   PF04257;\n",
    "#=GF MB   PF03976;\n",
    "#=GF MB   PF03796;\n",
    "#=GF MB   PF03567;\n",
    "#=GF MB   PF03354;\n",
    "#=GF MB   PF02689;\n",
    "#=GF MB   PF02606;\n",
    "#=GF MB   PF02572;\n",
    "#=GF MB   PF02499;\n",
    "#=GF MB   PF02399;\n",
    "#=GF MB   PF01268;\n",
    "#=GF MB   PF00350;\n",
    "#=GF MB   PF00225;\n",
    "#=GF MB   PF00063;\n",
    "#=GF MB   PF10662;\n",
    "#=GF MB   PF00271;\n",
    "#=GF MB   PF00910;\n",
    "#=GF MB   PF05496;\n",
    "#=GF MB   PF02562;\n",
    "#=GF MB   PF00025;\n",
    "#=GF MB   PF05049;\n",
    "#=GF MB   PF03266;\n",
    "#=GF MB   PF01591;\n",
    "#=GF MB   PF00071;\n",
    "#=GF MB   PF04548;\n",
    "#=GF MB   PF00009;\n",
    "#=GF MB   PF00176;\n",
    "#=GF MB   PF07517;\n",
    "#=GF MB   PF03237;\n",
    "#=GF MB   PF00735;\n",
    "#=GF MB   PF04665;\n",
    "#=GF MB   PF02263;\n",
    "#=GF MB   PF01926;\n",
    "#=GF MB   PF00580;\n",
    "#=GF MB   PF00270;\n",
    "#=GF MB   PF09848;\n",
    "#=GF MB   PF06745;\n",
    "#=GF MB   PF04851;\n",
    "#=GF MB   PF01443;\n",
    "#=GF MB   PF03193;\n",
    "#=GF MB   PF00503;\n",
    "#=GF MB   PF06858;\n",
    "#=GF MB   PF02283;\n",
    "#=GF MB   PF02456;\n",
    "#=GF MB   PF00154;\n",
    "#=GF MB   PF03029;\n",
    "#=GF MB   PF08477;\n",
    "#=GF MB   PF02421;\n",
    "#=GF MB   PF12696;\n",
    "#=GF MB   PF07999;\n",
    "#=GF MB   PF04310;\n",
    "#=GF MB   PF05272;\n",
    "#=GF MB   PF06048;\n",
    "#=GF MB   PF12774;\n",
    "#=GF MB   PF12775;\n",
    "#=GF MB   PF12780;\n",
    "#=GF MB   PF04317;\n",
    "#=GF MB   PF12846;\n",
    "#=GF MB   PF11398;\n",
    "#=GF MB   PF13086;\n",
    "#=GF MB   PF13087;\n",
    "#=GF MB   PF13166;\n",
    "#=GF MB   PF13173;\n",
    "#=GF MB   PF13175;\n",
    "#=GF MB   PF13177;\n",
    "#=GF MB   PF13189;\n",
    "#=GF MB   PF10923;\n",
    "#=GF MB   PF13191;\n",
    "#=GF MB   PF08303;\n",
    "#=GF MB   PF13207;\n",
    "#=GF MB   PF13238;\n",
    "#=GF MB   PF13245;\n",
    "#=GF MB   PF13304;\n",
    "#=GF MB   PF13307;\n",
    "#=GF MB   PF13361;\n",
    "#=GF MB   PF13401;\n",
    "#=GF MB   PF13469;\n",
    "#=GF MB   PF12128;\n",
    "#=GF MB   PF13479;\n",
    "#=GF MB   PF13476;\n",
    "#=GF MB   PF13481;\n",
    "#=GF MB   PF13500;\n",
    "#=GF MB   PF13514;\n",
    "#=GF MB   PF13521;\n",
    "#=GF MB   PF13538;\n",
    "#=GF MB   PF13555;\n",
    "#=GF MB   PF13558;\n",
    "#=GF MB   PF13604;\n",
    "#=GF MB   PF13614;\n",
    "#=GF MB   PF13654;\n",
    "#=GF MB   PF13671;\n",
    "#=GF MB   PF13872;\n",
    "#=GF MB   PF14516;\n",
    "#=GF MB   PF14532;\n",
    "#=GF MB   PF14617;\n",
    "#=GF MB   PF05625;\n",
    "#=GF MB   PF05179;\n",
    "#=GF MB   PF16203;\n",
    "#=GF MB   PF14417;\n",
    "#=GF MB   PF02500;\n",
    "#=GF MB   PF13337;\n",
    "#=GF MB   PF11602;\n",
    "#=GF MB   PF16575;\n",
    "#=GF MB   PF16796;\n",
    "#=GF MB   PF13871;\n",
    "#=GF MB   PF02702;\n",
    "#=GF MB   PF10088;\n",
    "#=GF MB   PF03846;\n",
    "#=GF MB   PF17213;\n",
    "#=GF MB   PF12848;\n",
    "#=GF MB   PF12399;\n",
    "#=GF MB   PF14396;\n",
    "#=GF MB   PF10996;\n",
    "#=GF MB   PF09711;\n",
    "#=GF MB   PF11111;\n",
    "#=GF MB   PF08351;\n",
    "#=GF MB   PF03192;\n",
    "#=GF MB   PF02841;\n",
    "#=GF MB   PF05609;\n",
    "#=GF MB   PF08438;\n",
    "#=GF MB   PF12344;\n",
    "#=GF MB   PF12781;\n",
    "#=GF MB   PF03028;\n",
    "#=GF MB   PF16813;\n",
    "#=GF MB   PF16834;\n",
    "#=GF MB   PF16836;\n",
    "#=GF MB   PF07034;\n",
    "#=GF MB   PF10483;\n",
    "#=GF MB   PF09807;\n",
    "#=GF MB   PF03618;\n",
    "#=GF MB   PF17784;\n",
    "#=GF MB   PF18128;\n",
    "#=GF MB   PF18133;\n",
    "#=GF MB   PF07529;\n",
    "#=GF MB   PF18747;\n",
    "#=GF MB   PF18748;\n",
    "#=GF MB   PF18751;\n",
    "#=GF MB   PF18766;\n",
    "#=GF MB   PF18082;\n",
    "#=GF MB   PF19044;\n",
    "#=GF MB   PF19263;\n",
    "//`\n",
    "\n",
    "    - Tubulin (CL0566)\n",
    "    \n",
    "    `\n",
    "    Tubulin\tPF00091\tDomain\t87\t23201\t177.40\t33\t42.70\t\tChanged\tTubulin/FtsZ family, GTPase domain\n",
    "    Tubulin-binding\tPF00418\tFamily\t87\t7623\t30.40\t52\t13.67\t\tChanged\tTau and MAP protein, tubulin-binding repeat\n",
    "    Tubulin_2\tPF13809\tFamily\t45\t515\t329.60\t19\t31.88\t\tChanged\tTubulin like\n",
    "    Tubulin_3\tPF14881\tDomain\t17\t1193\t183.10\t30\t33.19\t\tChanged\tTubulin domain\n",
    "    Tubulin_C\tPF03953\tDomain\t34\t12921\t123.40\t45\t28.10\t\tChanged\tTubulin C-terminal domain\n",
    "    `\n",
    "    \n",
    "    `\n",
    "    grep -A 50 'CL0566' Pfam-C\n",
    "#=GF AC   CL0566.2\n",
    "#=GF DE   Tubulin nucleotide-binding domain-like, GTPase\n",
    "#=GF AU   Coggill P\n",
    "#=GF RN   [1]\n",
    "#=GF RM   9628483\n",
    "#=GF RT   Tubulin and FtsZ form a distinct family of GTPases.\n",
    "#=GF RA   Nogales E, Downing KH, Amos LA, Lowe J;\n",
    "#=GF RL   Nat Struct Biol. 1998;5:451-458.\n",
    "#=GF RN   [2]\n",
    "#=GF RM   19697711\n",
    "#=GF RT   Tubulin and its prokaryotic homologue FtsZ: a structural and\n",
    "#=GF RT   functional comparison.\n",
    "#=GF RA   Dyer N;\n",
    "#=GF RL   Sci Prog. 2009;92:113-137.\n",
    "#=GF RN   [3]\n",
    "#=GF RM   16545939\n",
    "#=GF RT   An anthropoid-specific segmental duplication on human chromosome\n",
    "#=GF RT   1q22.\n",
    "#=GF RA   Kuryshev VY, Vorobyov E, Zink D, Schmitz J, Rozhdestvensky TS,\n",
    "#=GF RA   Munstermann E, Ernst U, Wellenreuther R, Moosmayer P, Bechtel S,\n",
    "#=GF RA   Schupp I, Horst J, Korn B, Poustka A, Wiemann S;\n",
    "#=GF RL   Genomics. 2006;88:143-151.\n",
    "#=GF DR   SCOP; 524907;\n",
    "#=GF DR   CATH; 3.40.50.1440;\n",
    "#=GF CC   The characteristic families of this clan are the tubulin GTPase\n",
    "#=GF CC   domain and the alpha and beta tubulin subunits. A segmental\n",
    "#=GF CC   duplication early on in the primate lineage has led to the\n",
    "#=GF CC   creation of up to three tubulin-like domains in the higher\n",
    "#=GF CC   primates. The most N-terminal one is the Misato domain, which\n",
    "#=GF CC   probably represents the original domain [3].\n",
    "#=GF MB   PF10644;\n",
    "#=GF MB   PF14881;\n",
    "#=GF MB   PF13809;\n",
    "#=GF MB   PF00091;\n",
    "//\n",
    "#=GF ID   Tubulin_C\n",
    "#=GF AC   CL0442.4\n",
    "#=GF PI   Tubulin_C; Tubulin-like;\n",
    "#=GF DE   Tubulin, FtsZ and Misato and their C-termini,\n",
    "#=GF AU   Coggill P\n",
    "#=GF DR   SCOP; 55307;\n",
    "#=GF DR   CATH; 3.30.1330.20;\n",
    "#=GF CC   This superfamily is characterised by being at the C-terminus of the\n",
    "#=GF CC   GTPase-binding region of tubulins.\n",
    "#=GF MB   PF03953;\n",
    "#=GF MB   PF12327;\n",
    "//\n",
    "`\n",
    "\n",
    "    - Actin_ATPase (CL0108)\n",
    "    \n",
    "    `\n",
    "    Actin\tPF00022\tFamily\t24\t22115\t325.00\t28\t82.77\t\tChanged\tActin\n",
    "    Actino_peptide\tPF14408\tFamily\t15\t124\t61.60\t34\t66.86\t\tChanged\tRibosomally synthesised peptide in actinomycetes\n",
    "    Actin_micro\tPF17003\tFamily\t7\t29\t267.20\t17\t71.54\t\tChanged\tPutative actin-like family\n",
    "    `\n",
    "\n",
    "    -\n",
    "    `\n",
    "    Hexokinase_1\tPF00349\tDomain\t387\t6723\t191.40\t36\t40.35\t\tChanged\tHexokinase\n",
    "    Hexokinase_2\tPF03727\tDomain\t420\t6961\t228.90\t34\t49.48\t\tChanged\tHexokinase\n",
    "    `\n",
    "    \n",
    "    `\n",
    "    grep -A 60 'CL0108' Pfam-C\n",
    "#=GF AC   CL0108.16\n",
    "#=GF DE   Actin-like ATPase Superfamily\n",
    "#=GF AU   Finn RD\n",
    "#=GF RN   [1]\n",
    "#=GF RM   15533433\n",
    "#=GF RT   The structure of the cytoplasmic domain of EpsL, an inner membrane\n",
    "#=GF RT   component of the type II secretion system of Vibrio cholerae: an\n",
    "#=GF RT   unusual member of the actin-like ATPase superfamily. \n",
    "#=GF RA   Abendroth J, Bagdasarian M, Sandkvist M, Hol WG; \n",
    "#=GF RL   J Mol Biol 2004;344:619-633.\n",
    "#=GF DR   CATH; 3.30.420.40;\n",
    "#=GF DR   CATH; 3.40.367.20;\n",
    "#=GF DR   SCOP; 53067;\n",
    "#=GF CC   The actin-like ATPase domain forms an alpha/beta canonical fold. The\n",
    "#=GF CC   domain can be subdivided into 1A, 1B, 2A and 2B subdomains.\n",
    "#=GF CC   Subdomains 1A and 1B share the same RNAseH-like fold (a five-stranded\n",
    "#=GF CC   beta-sheet decorated by a number of alpha-helices). Domains 1A and 2A\n",
    "#=GF CC   are conserved in all members of this superfamily, whereas domain 1B\n",
    "#=GF CC   and 2B have a variable structure and are even missing from some\n",
    "#=GF CC   homologues [1].  Within the actin-like ATPase domain the ATP-binding\n",
    "#=GF CC   site is highly conserved. The phosphate part of the ATP is bound in a\n",
    "#=GF CC   cleft between subdomains 1A and 2A, whereas the adenosine moiety is\n",
    "#=GF CC   bound to residues from domains 2A and 2B[1].\n",
    "#=GF MB   PF06406;\n",
    "#=GF MB   PF00480;\n",
    "#=GF MB   PF02541;\n",
    "#=GF MB   PF00814;\n",
    "#=GF MB   PF06723;\n",
    "#=GF MB   PF05378;\n",
    "#=GF MB   PF01968;\n",
    "#=GF MB   PF00012;\n",
    "#=GF MB   PF03727;\n",
    "#=GF MB   PF00349;\n",
    "#=GF MB   PF02685;\n",
    "#=GF MB   PF01150;\n",
    "#=GF MB   PF03630;\n",
    "#=GF MB   PF00370;\n",
    "#=GF MB   PF02782;\n",
    "#=GF MB   PF06277;\n",
    "#=GF MB   PF02543;\n",
    "#=GF MB   PF03309;\n",
    "#=GF MB   PF01869;\n",
    "#=GF MB   PF00022;\n",
    "#=GF MB   PF00871;\n",
    "#=GF MB   PF03702;\n",
    "#=GF MB   PF08841;\n",
    "#=GF MB   PF07318;\n",
    "#=GF MB   PF05134;\n",
    "#=GF MB   PF11104;\n",
    "#=GF MB   PF13941;\n",
    "#=GF MB   PF14450;\n",
    "#=GF MB   PF09989;\n",
    "#=GF MB   PF06050;\n",
    "#=GF MB   PF17003;\n",
    "#=GF MB   PF14574;\n",
    "#=GF MB   PF17788;\n",
    "#=GF MB   PF17989;\n",
    "//`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201002\n",
    "- Further grabbing datasets, details in data_prep_200930\n",
    "    -TODO: actin-like and p_loop_gtpase mapping to uniref50 sequences\n",
    "\n",
    "- bi-directional mLSTM building up\n",
    "    - adopt the same embedding layer as Will's model\n",
    "    - 10 dimension amino acid embedding + 128 hidden states (64 for each forward and backward)\n",
    "- GRU on a larger dataset:\n",
    "    - Use the dev, train, and test from kaggle challenge as general protein source\n",
    "    - alternative, may use uniref50_01\n",
    "    - After finishing training, embed cymobase/4clans\n",
    "    - Alternatively, then fine_tune the model with 4clans fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201003\n",
    "- Training Will's GRU model on 1200k set from kaggle\n",
    "    - TO-DO: \n",
    "    1. embed the 4 pfam clans sequence to so the groupings of families\n",
    "    2. further evotune the model using 4 pfam clans\n",
    "    3. evotune potential using a class prediction task?\n",
    "\n",
    "- Bi-directional LSTM first test on ~1k uniref50/ ~1k kaggle domain sequences to make sure it works\n",
    "    - add dropout layers\n",
    "    - apply next token prediction \n",
    "    - change it to a domain classification task"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201009\n",
    "- build Seq2Seq with attention for next token/masked token prediction\n",
    "    - https://arxiv.org/abs/1506.03134\n",
    "    - https://github.com/spro/practical-pytorch/blob/master/seq2seq-translation/seq2seq-translation.ipynb\n",
    "    - https://github.com/ast0414/pointer-networks-pytorch/blob/master/train_sort.py\n",
    "- build transformer model and use the encoder to generate embedding \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201012\n",
    "## 1. Compile data for embedding visualization\n",
    "### Target1: Inspect the embedding on 5 types in motor toolkits []\n",
    "- [https://link.springer.com/referenceworkentry/10.1007%2F978-3-642-16712-6_762]\n",
    "- [https://www.sciencedirect.com/science/article/pii/S0092867403001119]\n",
    "    1. Conventional Kinesin: KIF5B, KHC, NKin, KLP1, KinA, DdK5\n",
    "    2. Kinesin II: 1) heterodimer 2) homodimer: Osm3/KIF17 three Kinesin-2 subfamilies exist: the KIF3A, KIF3B/C, and KIF17 subfamilies\n",
    "    3. Kinesin III: KIF1A, KIF1B, KIF13A, UNC104, DdUnc104\n",
    "    4. Cytoplasmic dynein\n",
    "    5. myosin V\n",
    "\n",
    "### Target II: Inspect the embedding on pfam family of molecular motors of interests\n",
    "    1. actin-like: \n",
    "        - PF00349\tHexokinase_1\t2704\tactin_like\n",
    "        - PF00022\tActin\t7689\tactin_like\n",
    "        - PF03727\tHexokinase_2\t2859\tactin_like\n",
    "        - PF06723\tMreB_Mbl\t3621\tactin_like\n",
    "        - PF14450\tFtsA\t4161\tactin_like\n",
    "    2. tubulin_c\n",
    "        - PF03953\tTubulin_C\t1999\ttubulin_c\n",
    "        - PF12327\tFtsZ_C\t2756\ttubulin_c\n",
    "    3. tubulin_binding \n",
    "        - PF00091\tTubulin\t5951\ttubulin_binding\n",
    "        - PF10644\tMisat_Tub_SegII\t628\ttubulin_binding\n",
    "        - PF13809\tTubulin_2\t396\ttubulin_binding\n",
    "        - PF14881\tTubulin_3\t660\ttubulin_binding\n",
    "    4. p_loop_gtpase\n",
    "        - PF00063\tMyosin_head\t8304\tp_loop_gtpase\n",
    "        - PF00225\tKinesin\t15099\tp_loop_gtpase\n",
    "        - PF03028\tDynein_heavy\t2322\tp_loop_gtpase\n",
    "\n",
    "###  Produce the background embedding space \n",
    "    - background 1: target VS all pfam motors (from superfamily)\n",
    "    - background 2: target VS the pfam superfamily they belongs to\n",
    "    - background 3: target VS randomly selected pfam sequences across all families\n",
    "    - background 4: different types of target 1 in their respective pfam family/superfamily background\n",
    "\n",
    "## 2. Generate Emdedding for the data\n",
    "    - Seq2Seq small, LSTM_5(on kaggle pfam) and transformer (on kaggle pfam) are currently available"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201021\n",
    "### TO-DO:\n",
    "    1. More balanced training dataset\n",
    "    2. Evo-tuning on motor proteins\n",
    "    3. Directly train on motor proteins\n",
    "    4. Supervised learning (on kinesin type, for example)\n",
    "    5. Fine tune training hyperparameters (for transformer especially)\n",
    "    6. More balanced dataset to be visualized\n",
    "    7. Other dimensionality reduction methods\n",
    "    8. Testing on proteins with phylogenetic relation/protein at different stability level\n",
    "\n",
    "### Some quick stuffs to start with\n",
    "    0. Inspect the SVD variance of each dimension in transformer example, and result embedding/encodering hidden state amount if too many dimensions are useless\n",
    "    \n",
    "    1. Evo-tune the current model \n",
    "        - \" To perform evolutionary fine-tuning (which we call ‘evotuning’), we ran ~13,000 unsupervised weight updates of the UniRep mLSTM (Evotuned UniRep) performing the same next-character prediction task on a set of ~25,000 likely evolutionarily related sequences obtained via JackHMMER search (Methods). We compared this to the untuned, global UniRep as well as a randomly initialized UniRep architecture trained only on local evolutionary data\"\n",
    "        - \"Evotuned UniRep demonstrated superior performance in both tasks, having near ideal recall for small (<30 sequences) testing budgets and quickly recovering the brightest sequence in the generalization set of >33,000 sequences with fewer than 100 sequences tested (Fig. 4d, red). The global information captured by UniRep was critical for this gain as evotuning from a random initialization of the same architecture yielded inferior performance\"\n",
    "        - for now, use pfam_motor_balanced + pfam_motor_target dataset to evo-tune, since the prior captures either a very balanced set of proteins from motor pfam, and the latter captures a variety of pfams of most interests. \n",
    "        - **Shuffle to avoid contiguity**\n",
    "        - 3 models + 2 dataset --> 6 models\n",
    "     \n",
    "     2. Directly training on pfam motors ---> shuffle it as the same pfams are grouped together\n",
    "     \n",
    "     3. Create visualization dataset on original model but with distantly related pfams\n",
    "         - a prokaryote pfam, motors and some randomly selected ones\n",
    "         - reproduce the one from Will's GRU example\n",
    "         - find a well-known phylogenetically related one\n",
    "         - find a stability related set of protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201101\n",
    "\n",
    "0. Output PCA/TSNE for facebook embedding on the datasets\n",
    "\n",
    "1. Inspect and summarize the performance of LSTM5, Seq2Seq, TransformerEncoder_causal, TransformerEncoder_mlm, \n",
    "\n",
    "2. Edit the Masked language model to add mask and start of sentence token into the vocabulary, the train without mask, Use Uniref50 to train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201110 Group Meeting Summary\n",
    "\n",
    "### Based on the current work\n",
    "- Seems that ESM model without evotuning is the best embedding model so far that capture both phylogenetic and functional/structural similarity.\n",
    "- standardize hidden state vectors before TSNE\n",
    "- Inspect the overlapping region of the TSNE plot and see whether the overlapping proteins are of interesting organism\n",
    "\n",
    "### Diving deeper into the methods\n",
    "- Carefully walk through the protein engineering approach/ application side of UniRep, ESM, and ProGen, and 3D\n",
    "- Construct some structurally/functionally related testing dataset both including motor related and other stuffs\n",
    "- Try explore the 3D convolution of kinesin/tubulin\n",
    "- concat embedding hidden state and 3D convolutional features as a new vector and then perform\n",
    "\n",
    "### Biological data grabbing\n",
    "- Learn try to train a model with organisms temperature as label and protein sequence as input data (compile motor proteins with interesting features.)\n",
    "- GFP protein engineering data and thermostability data, as well as CAGI/CASP data on protein stability\n",
    "- Compile more kinesin data (kinesin sub-families that have different properties)\n",
    "- Look into the APTase domain of the kinesin\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201111\n",
    "1. Normalize the hidden states of ESM models and mlm models, then perform the pca/tsne again to see if the high dependency on PC1 disappear and if any of the new trends appear\n",
    "    - Obersvation: From SVD analysis, the variance accounted by each PCs did get a more even distribution (with the order of magnitude difference between x4), and the later PCs has a more even variance being accounted. \n",
    "    - From the plot: The clusters are still clear and at some point seem to group closer together. For t34, the linear line trend still remain but got questionably mitigated. Overall, normalization helped little and may let the hidden state lose scaling information, and so some families are no longer grouped into a single group (which could be interesting if we would like to study the intersection of the proteins.)\n",
    "\n",
    "2. Read Carefully through ESM paper, take note about the dataset and engineering side they applied:\n",
    "    - "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 201115 \n",
    "\n",
    "## Generate UMAP embedding for transformer model t12 and t34\n",
    "- code in \n",
    "\n",
    "## Preperation of temperature/thermostability/mutant stability    \n",
    "- Protein thermostability engineering [https://pubs.rsc.org/en/content/articlehtml/2016/ra/c6ra16992a]\n",
    "\n",
    "### - TEMPURA: Database of growth TEMPeratures of Usual and RAre prokaryotes [http://togodb.org/db/tempura]\n",
    "### - A Method for Prediction of Thermophilic Protein Based on Reduced Amino Acids and Mixed Features [https://www.frontiersin.org/articles/10.3389/fbioe.2020.00285/full]\n",
    "    - The temperature of thermophilic proteins in this dataset was set to above 60°C and the temperature of non-thermophilic proteins was set to be less than 30°C. After removing redundancy and homology bias, there were 915 thermophilic and 793 non-thermophilic proteins. These data can be downloaded from http://www.labio.info/index-1therm.html.\n",
    "    - Thus, the 20 amino acids were classified into 19 types in the above scheme (Susko and Roger, 2007), in which Ile (I) and Val (V) were viewed as a single type, while every one of other categories had only one amino acid.\n",
    "\n",
    "## from https://github.com/magratheaner/igem-insilico-wiki#by-dataset\n",
    "### - ProtDataTherm: A database for thermostability analysis and engineering of proteins [https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5788348/#!po=50.0000]\n",
    "    - the ProtDataTherm dataset claims to offer ~600k thermostable (Tm >40°C) proteins categorized by domain families (Pfams). Although it is badly cleaned and only available via an intransparent web interface\n",
    "    - [https://drive.google.com/drive/folders/1oOkUYSJXy8NUP1sQ5U0GQFNvSuwdjU67]\n",
    "\n",
    "### - istable2.0: Training and testing dataset\n",
    "\n",
    "### - HotMuSiC v2.0 with sequence\n",
    "    - https://drive.google.com/file/d/1-4-wFMtTZKfVBLcvQtVcCd26NtgXFmmc/view\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201118\n",
    "1. Install TAPE and output motor protein embeddings for bert-base and unirep babbler_1900 and esm\n",
    "2. Help with David to collect unicellular organism kenisin and tubulin full sequence and taxonomy from uniprot\n",
    "3. Classification Model (SVM,RF,Logistic Regression,FNN)\n",
    "4. Use the above pre-trained classifier to train seqGAN/GAN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201123\n",
    "- Read through paper that Matt and other members recommended\n",
    "- Compile Kinesin dataset from the following 2 dataset\n",
    "    - https://sites.duke.edu/kinesin/\n",
    "    - http://cb.m.u-tokyo.ac.jp/KIF/\n",
    "- For each entry, try query the full uniprot sequence and other information\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201130\n",
    "- Clean up the 2 Kinesin dataset and provide visualization/stats\n",
    "- Kinesin-14 motors display a slow velocity towards the microtubule minus-end compared with the plus-end-directed conventional kinesin (42.7±2.19 nm/s) \n",
    "\n",
    "### Output table kif_uniprot.csv\n",
    "- kif uniprot contains 2 set of uniprot sequences entry:\n",
    "- kif japan table contains protein accession and was used to directly query uniprot\n",
    "- kif duke: for the entries with documentation on speed and direction, blast against swissprot and get all the accessions corresponding to each entry, then the uniprot accession was used to query uniprot sequence and further information we could query the sequence by first selecting an entry in kif_doc, then use the entry stored to query sequence\n",
    "\n",
    "### Output table kif_speed.csv\n",
    "- contain information about the sequences that has well-documented speed and direction information, and the uniprot accession of sequences shown in as a match when blasting against the swissprot database\n",
    "\n",
    "### Motor toolkit contains kinesin, dyeinin and myosin V sequences based on regex expression search\n",
    "- for now limit it down only to kinesin and dyenin, as myosin V showed misleading results\n",
    "- also train one with only the kenisin\n",
    "\n",
    "### Perform weight updates using kinesin+dyeinin/kinesin only from the kif_uniprot+motor_toolkit/kinesin only curated from kif_uniprot dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201204\n",
    "\n",
    "### Compile kinesin/dynein from different sources\n",
    "  - compile sources of different kinesin sequences and make a large table for all the documentation\n",
    "  - A uniprot sequence table should contain uniprot_acc|seq|type, where type indicate whether the sequence is for kinesin or dynein\n",
    "  - The documentation on speed (from Duke's website) link to the accessions related to kif identifier\n",
    "  - The documentation on kinesin type for each uniprot accession (merging kinesin_labelled,kif_jp and kif_duke)\n",
    "  >   - 0: unlabelled kinesin family\n",
    "      - 1~14: the labelled kinesin families\n",
    "      - 15: Orphan\n",
    "      - 16: Kinesin Light Chain (KLC)\n",
    "      \n",
    "`      \n",
    "          Entry\tdb_acc\n",
    "db_name\t\t\n",
    "kif_duke\t85\t85\n",
    "kif_jp\t456\t456\n",
    "motor_toolkit\t57\t57\n",
    "`\n",
    "\n",
    "### Visualization of thermostability prediction\n",
    "  1. the label in a embedded space\n",
    "  2. the correctly predicted VS incorrectly prediced entries in the embedded space\n",
    "  3. prediction/real group by pfam family/clan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201207\n",
    "- embed the kinesins and conduct a direction/family visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 201211\n",
    "- Sample a subset from the kif embeddings for each family and perform an even embedding using umap/tsme for each of \n",
    "    - to ensure that the family annotation is consistent and the most reliable, use the entries from kif_jp for subset\n",
    "- If the grouped sampling is no any better, try first perform PCA then tsne/umap\n",
    "- normalized/unnormalized + subset/non-subset + tsne/pca/umap generation for the kif dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
